File: apps/api/src/solidchain/schemas/vectorstore.py
from enum import Enum
from typing import List

from pydantic import BaseModel

from solidchain.schemas.embeddings import Embeddings
from solidchain.schemas.mixins.dates import TimestampMixin
from solidchain.schemas.storage import FileStorage


class VectorStoreDB(str, Enum):
    CHROMA = "chroma"
    ELASTIC_SEARCH = "elastic_search"
    FAISS = "faiss"
    MILVUS = "milvus"
    PINECONE = "pinecone"
    QDRANT = "qdrant"
    WEAVIATE = "weaviate"


# Shared properties
class VectorStoreBase(BaseModel):
    vectorstoreId: int
    # userId: int
    name: str
    description: str
    vectorDb: str
    embeddingsType: Embeddings
    urls: List[str]
    files: List[FileStorage]
    index: FileStorage


# Properties shared by models stored in DB
class VectorStoreBaseInDBBase(VectorStoreBase, TimestampMixin):
    pass

    class Config:
        orm_mode = True


# Properties to return to client
class VectorStore(VectorStoreBaseInDBBase):
    pass


# Properties properties stored in DB
class VectorStoreBaseInDBBaseInDB(VectorStoreBaseInDBBase):
    pass

File: apps/web/src/lib/validators/VectorStore.ts
import { z } from "zod";
import { fileStorageValidator } from "~/lib/validators/FileStorage";
import { timestampValidator } from "~/lib/validators/Timestamp";
import { embeddingsValidator } from "./Embeddings";

export const vectorDbValidator = z.enum([
    "chroma",
    "elastic_search",
    "faiss",
    "milvus",
    "pinecone",
    "qdrant",
    "weaviate"
]);
export type VectorDbType = z.infer<typeof vectorDbValidator>;

export const vectorStoreValidator = z
    .object({
        vectorstoreId: z.number(),
        name: z.string(),
        description: z.string(),
        vectorDb: vectorDbValidator,
        embeddingsType: embeddingsValidator,
        files: z.array(fileStorageValidator),
        index: fileStorageValidator.nullish()
    })
    .merge(timestampValidator);
export type VectorStore = z.infer<typeof vectorStoreValidator>;
